prototype

stefanfoulis opened the pull request:
- [x] basic setup and design decisions
- [x] basic aldryn-addons app
- [x] basic aldryn-django Addon
- [x] prototype docs and guide
- [x] holidays until 24.8.2015 [:arrows_counterclockwise: @stefanfoulis]
- [x] basic aldryn-cms Addon
- [x] ~~decide where to put initial aldryn-addons config and dependency (chicken/egg problem)~~ add itself to INSTALLED_APPS when loading settings from addons
- [x] decide where to put aldryn-addons dependency (chicken/egg problem)
- [x] pip-tools / pip-compile logic
- [x] utility function to get variable from env, then settings, then default
## aldryn-django
- [x] manage.py api
- [x] wsgi.py api (including dj-static)
- [x] dynamic .env file parsing (useful for local development)
- [x] django.contrib.admin
- [x] database settings parsing
- [x] minimal default settings
- [x] Addon url api
- [x] templates
- [x] Addon migration api
- [x] collectstatic on build and at runtime
- [ ] env var to control whether templates are cached or not
- [x] timezone
- [x] uwsgi
- [x] re-integrate with launch scripts (pagespeed et al)
- [x] logging
- [x] logging: structlog support
- [x] cache
- [ ] email
- [ ] all sensible settings from the old base project
  - [ ] middleware/templates/context/....
  - [ ] ...
## aldryn-cms

will be split into sub-Addons later
- [x] basic cms
- [x] i18n (language) settings
- [x] parler
- [x] custom admin styling
### django-filer
- [x] basic
- [x] cms plugins
- [x] Pillow and easy-thumbnails
### various tools
- [ ] compress
- [x] sekizai
- [x] aldryn-boilerplates
- [ ] aldryn-sites
- [ ] aldryn-snake

stefanfoulis committed:
WIP

stefanfoulis committed:
better urls api

Chive commented on addons/aldryn-django/aldryn_config.py:
```diff
@@ -0,0 +1,84 @@
+        settings['DATA_ROOT'] = env('DATA_ROOT', os.path.join(settings['BASE_DIR'], 'data'))
+        settings['SECRET_KEY'] = env('SECRET_KEY', 'this-is-not-very-random')
+        settings['DEBUG'] = boolean_ish(env('DEBUG', False))
+        settings['TEMPLATE_DEBUG'] = boolean_ish(env('DEBUG', settings['DEBUG']))
```
what about

```
settings['TEMPLATE_DEBUG'] = boolean_ish(env('TEMPLATE_DEBUG', settings['DEBUG']))
```

Chive commented on addons/aldryn-django/aldryn_config.py:
```diff
@@ -0,0 +1,84 @@
+            )
+        settings['DATABASES']['default'] = dj_database_url.parse(settings['DATABASE_URL'])
+
+        settings['ALLOWED_HOSTS'] = env('ALLOWED_HOSTS', ['*'])
```
let's use `ALLOWED_HOSTS = ['localhost', '*']` instead. I had issues before where Django checked `if 'localhost' in ALLOWED_HOSTS` and failed because of that

Chive commented on addons/aldryn-django/aldryn_django/i18n_urls.py:
```diff
@@ -0,0 +1,15 @@
+    '',
+    url(r'^admin/', include(admin.site.urls)),
+)
+print 'patterns: ', urlpatterns
```
`print`

Chive commented on addons/aldryn-django/aldryn_django/startup.py:
```diff
@@ -0,0 +1,33 @@
+    _setup(path=path)
+    from django.core.wsgi import get_wsgi_application
+    from dj_static import Cling
+    # evil workaround for gunicorn not being able to import everything
```
uwsgi instead?

stefanfoulis replied:
yeah, I'm going to set it up with uwsgi. But it is likely that the same hack is needed for uwsgi as well.

Chive commented on addons/aldryn-django/setup.py:
```diff
@@ -0,0 +1,27 @@
+        'raven',
+        'opbeat',
+        # wsgi server related
+        'uwsgi',
```
above gunicorn, here uwsgi? i'm lost

Chive commented on aldryn-addons/aldryn_addons/settings.py:
```diff
@@ -0,0 +1,106 @@
+    shutil.rmtree(debug_path, ignore_errors=True)
+    os.makedirs(debug_path)
+
+    def dump(obj, count, name):
```
very hard to read, better:

``` python
def dump(obj, count, name):
    dump_name = '{}-{}.dump'.format(count_str(count), name)
    dump_path = os.path.join(debug_path, dump_name)
    save_settings_dump(obj, dump_path)
    return count + 1
```

Chive commented on aldryn-addons/aldryn_addons/urls.py:
```diff
@@ -0,0 +1,27 @@
+def patterns():
+    args = []
+    for pat in getattr(settings, 'ADDON_URLS', []):
+        print pat
```
`print`?

stefanfoulis replied:
work in progress ;-)

Chive commented on aldryn-addons/aldryn_addons/urls.py:
```diff
@@ -0,0 +1,27 @@
+    urlpatterns = urls.patterns('', *args)
+    args = []
+    for pat in getattr(settings, 'ADDON_URLS_I18N', []):
+        print pat
```
`print`?

Chive commented on aldryn-addons/aldryn_addons/urls.py:
```diff
@@ -0,0 +1,27 @@
+from django.conf import urls
+from django.conf.urls.i18n import i18n_patterns
+
+def patterns():
```
```python
def patterns():
    args = []
    for pat in getattr(settings, 'ADDON_URLS', []):
        args.append(urls.url(r'^', urls.include(pat)))
    urlpatterns = urls.patterns('', *args)
    args = []
    for pat in getattr(settings, 'ADDON_URLS_I18N', []):
        args.append(urls.url(r'^', urls.include(pat)))
    i18n_urlpatterns = i18n_patterns('', *args)
    return urlpatterns + i18n_urlpatterns
```

stefanfoulis replied:
I actually refactored this to be way simpler.

Chive commented:
Very good.

stefanfoulis committed:
fixes

stefanfoulis committed:
boilerplate enhancements

stefanfoulis committed:
basic logging setup

stefanfoulis committed:
add Procfile

stefanfoulis committed:
add cache configuration

stefanfoulis committed:
aldryn-django command with nginx/pagespeed config

aldryn-django command to invoce the webserver. It encapsulates the logic
about deciding which mode to use (uwsgi only or nginx+uswgi) and
dynamically creates the nginx config for pagespeed.

stefanfoulis committed:
fix Procfile

stefanfoulis committed:
minimal i18n and timezone settings

stefanfoulis committed:
seperate addons-dev folder

stefanfoulis committed:
introduction docs

stefanfoulis committed:
remove unused stuff in Dockerfile

stefanfoulis committed:
update to new project name

stefanfoulis committed:
Update README.md

stefanfoulis committed:
wording change

stefanfoulis committed:
some restructuring

stefanfoulis committed:
fix link to meta project

jqb commented on README.md:
````diff
@@ -1,3 +1,131 @@
+lifting is outsourced into Addons.
+
+```
+addons   # all Addons installed on this website
````
So... it's actually in `.gitignore` too, right?

stefanfoulis replied:
No. This would contain the following files in git version history:
- `addon.json`: copied in by the controlpanl each time the addon is updated
- `aldryn_config.py`: copied in by the controlpanl each time the addon is updated
- `settings.json`: all **non-secret** settings of the addon (will automatically be updated by the controlpanel when changes are made over the UI)

In a later iteration of the api, we could have addon authors put `aldryn_config.py` inside their python package. Then it would not have to be copied into this directory and could be read directly from the installed package from site-packages. `addon.json` could probably be made obsolete, since the only information it holds it the "package name" used on aldryn. And that is identical to the pypi package name in 99% of the cases.

jqb commented on README.md:
```diff
@@ -1,3 +1,131 @@
+migrations and fix-mptt by django-cms).
+So upgrading the common ``aldryn-django`` package can also
+upgrade the way django is started and how migrations are run (Django 1.6 to
+Django 1.7 migrations upgrade anyone? ;-) ).
```
If even django is an addon and, as far as I can see, it provides `aldryn-django` command - why do we even need `manage.py`? Historical reasons?

stefanfoulis replied:
`manage.py` is there for familiarity, so a django developer feels at home right away. But true, we could provide an `aldryn-django manage` command and just call that from this manage.py file.

And we could probably remove the `wsgi.py`, since it is usually not used by the developer directly.

jqb commented on addons-dev/aldryn-addons/aldryn_addons/settings.py:
```diff
@@ -0,0 +1,123 @@
+    debug_count = 0
+    debug_count = dump(settings, debug_count, 'initial')
+    # load global defaults
+    from django.conf import global_settings
```
Just curios... can't the import be in the beginning of the file?

stefanfoulis replied:
This code is run before/while django settings are configured. So to avoid circular imports and loading anything from django that requires django to already be configured, I tend to make imports from django as lazy as possible. For this specific import it could be put at the top though.

I also kind of like it to be together with the code that actually uses it in this case though (cleaner diff and related code together). What do you thinkg @jqb ? put it at the top anyway?

jqb commented on addons-dev/aldryn-addons/aldryn_addons/utils.py:
```diff
@@ -0,0 +1,37 @@
+def mkdirs(path):
+    try:
+        os.makedirs(path)
+    except:
```
Hm.... whenever I see `except` without exception type I'm suspicious. I'm kinda picky, but ... I'd put this `os.error` here (https://docs.python.org/2/library/os.html)

czpython replied:
If required.. then please use logging.

mikek replied:
I am not sure how/if the logging is configured on the aldryn platform, but if this is easily accessible by the user – yes, would be handy to see messages there.

stefanfoulis replied:
@czpython what would you like to log here? If the path already exists, there is nothing to do and all is fine (no log required). And if there is an exception, it is serious... so aborting completely with an exception should be ok.

jqb commented on addons-dev/aldryn-django/aldryn_django/cli.py:
```diff
@@ -0,0 +1,113 @@
+        raise click.UsageError('make sure you are in the same directory as manage.py')
+    from . import startup
+    startup._setup(BASE_DIR)
+    # TODO: there must be a better way
```
I think too. How about `pass_context` (http://click.pocoo.org/5/api/#click.pass_context) ?

jqb commented on addons-dev/aldryn-django/aldryn_django/models.py:
```diff
@@ -0,0 +1,3 @@
+#-*- coding: utf-8 -*-
+from __future__ import unicode_literals, absolute_import
+
```
What this import is for?

jqb commented on addons-dev/aldryn-addons/aldryn_addons/urls.py:
```diff
@@ -0,0 +1,16 @@
+    return [
+        urls.url(r'^', urls.include(url))
+        for url in getattr(settings, 'ADDON_URLS_I18N', [])
+    ]
```
A bit misleading format for `urls.py`. Should be probably indicated at least by some comments.

jqb commented on addons-dev/aldryn-django/aldryn_django/cli.py:
```diff
@@ -0,0 +1,113 @@
+    os.execvp(command, args)
+
+
+def start_uwsgi_command(port=None):
```
eh... passing settings as an arg is better idea. `port` is read from settings anyway.

stefanfoulis replied:
port is actually not the same all the time. By default it runs on `settings.PORT`, but if it is running behind nginx, nginx runs on `settings.PORT` and django runs on `settings.BACKEND_PORT`.
Port needs to be selectively overridable here: https://github.com/aldryn/prototype-aldryn-project-template/pull/1/files#diff-aa5fb7ca5f81e7999c77f439e08e7107R100

czpython commented on addons-dev/aldryn-addons/aldryn_addons/settings.py:
```diff
@@ -0,0 +1,123 @@
+    utils.mkdirs(settings['ADDONS_DIR'])
+    # TODO: .debug is not multi-process safe!
+    debug_path = os.path.join(settings['ADDONS_DIR'], '.debug')
+    shutil.rmtree(debug_path, ignore_errors=True)
```
is it necessary to remove everything in .debug every time this is called? Not sure how often it gets called so I ask :)

stefanfoulis replied:
It is run on every process startup (first time `django.conf.settings` is imported). Deleting probably causes problems with multiple processes start at the same time in the same environment. I'll probably deactivate it completely for the first release.

The data it outputs is pretty useful for debugging addon settings though. So I'm reluctant to remove it.

czpython commented on addons-dev/aldryn-addons/aldryn_addons/settings.py:
```diff
@@ -0,0 +1,123 @@
+    utils.mkdirs(debug_path)
+
+    def dump(obj, count, name):
+        dump_name = '{}-{}.dump'.format(count_str(count), name)
```
I assume python 2.6 is not supported

stefanfoulis replied:
nope :-)

czpython commented on addons-dev/aldryn-addons/aldryn_addons/settings.py:
```diff
@@ -0,0 +1,123 @@
+    from django.conf import global_settings
+    for key, value in global_settings.__dict__.items():
+        if not key in settings:
+            settings[key] = value
```
maybe `settings.setdefault(key, value)` instead of `if` ?

Chive commented:
Looks very good, awesome job @stefanfoulis! One thing I wanna tackle though and it might make sense to do it in the scope of this Epic is the docker image - or more its size. The images for websites are usually around 1 to 2 gigabytes - that's just too big. So it would make sense to look into where we can do improvements - we will benefit from shorter push/pulling times, more storage space and in general improved performance. Both me & @GaretJax have made good experiences with the tiny `Alpine` image so that might ba good fit.

mikek commented on addons-dev/aldryn-addons/aldryn_addons/utils.py:
```diff
@@ -0,0 +1,37 @@
+        value = value.lower()
+    if value in [True, 'true', '1', 'on', 'yes']:
+        return True
+    elif value in [False, 'false', '0', 'off', 'no']:
```
I'd add `'y'`/`'n'` as some commonly seen values.

mikek commented on addons-dev/aldryn-cms/aldryn_config.py:
```diff
@@ -0,0 +1,57 @@
+    name = forms.CheckboxField('Name', required=False)
+
+    def to_settings(self, data, settings):
+        # TODO: break out a lot of this stuff into other Addons
```
Are you going to add apphook-reload here? Or is this functionality supported on another level by the platform?

stefanfoulis replied:
We have an apphook-reload functionality already built into the platform.

mikek commented on addons-dev/aldryn-cms/setup.py:
```diff
@@ -0,0 +1,27 @@
+        'django-robots',
+        'aldryn-boilerplates',
+        'aldryn-sites',
+    ),
```
No strict versions for any of these?

stefanfoulis replied:
Stricter versions will be added here. But they can't be too strict, because dependency checking will catch impossible dependencies.

mikek commented on addons-dev/aldryn-django/aldryn_django/startup.py:
```diff
@@ -0,0 +1,26 @@
+    os.environ['DJANGO_SETTINGS_MODULE'] = env('DJANGO_SETTINGS_MODULE', 'settings')
+    dotenv_path = os.path.join(path, '.env')
+    if os.path.exists(dotenv_path):
+        print 'reading environment variables from {0}'.format(dotenv_path)
```
If logging is not available at this moment, can we use `sys.stdout.write()`? Is'nt it better than `print` statements in python 2.x style?

mikek commented on addons-dev/aldryn-django/aldryn_django/templates/aldryn_django/configuration/nginx.conf:
```diff
@@ -0,0 +1,61 @@
+  gzip on;
+  gzip_vary on;
+  gzip_comp_level 5;
+  gzip_min_length 10;
```
AFAIK this is too low and will just add more overhead. 1K can be a better starting point.

mikek commented on README.md:
```diff
@@ -1,3 +1,131 @@
+But you can do whatever you want in here. Including shooting yourself in the
+foot. The controlpanel will auto update the ``INSTALLED_ADDONS`` setting based on what
+is done in the UI.
+
```
I wonder if there is a good way to show user what standard Django settings values do we use by default and what override with this project template. So that he knows if it is safe/supported to set/override something.

Also, is there some plan to support project template updating (for fixes, for example)?

stefanfoulis committed:
many fixes based on feedback

also includes smarter env var and settings parsing

stefanfoulis committed:
migrate api

stefanfoulis committed:
add last url setting

stefanfoulis committed:
bare minimum django-cms working

stefanfoulis committed:
i18n settings, media/static serving, filer

stefanfoulis committed:
add aldryn-snake

stefanfoulis committed:
default cms plugins

filer
googlemap
link
snippet
text ck editor

also: select2 and captcha

stefanfoulis committed:
add ssl related packages

stefanfoulis committed:
don't ignore settings.json in addons-dev

if you have to code inline, you do want it.
if you have a subrepo, the subrepo needs to ignore it.

stefanfoulis committed:
add s3 and djeesefs storage configuration

stefanfoulis committed:
add django-compressor

stefanfoulis committed:
cmscloud basic functionality and aldryn-sites

stefanfoulis committed:
forwards compatible i18n_patterns signature

stefanfoulis committed:
changes based on feedback

stefanfoulis committed:
revert to not using pip-tools

because pip tools can't install from a tar.gz and we currently need that for private Addons

stefanfoulis committed:
simpler requirements.txt

stefanfoulis committed:
cleanup

stefanfoulis committed:
ignore node_modules

stefanfoulis committed:
add DOCKER_BUILD section in Dockerfile

for the future :-)

stefanfoulis commented:
this has been finished and moved to the real repo at aldryn/project-template
